internal/profiling/profiling.go

```diff
@@ -0,0 +1,213 @@
+// reference the entire lifetime of an RPC request, with Timers within it
+// representing different components such as encoding, compression, and
+// transport.
+//
+// The user is expected to use the included helper functions to do operations
+// on the Stat such as creating or appending a new timer. Direct operations on
+// the Stat's exported fields (which are exported for encoding reasons) may
+// lead to data races.
+type Stat struct {
+	StatTag  string
+	Timers   []Timer
+	Metadata []byte
+	mu       sync.Mutex
+}
+
+// A power of two that's large enough to hold all timers within an average RPC
+// request (defined to be a unary request) without any reallocation.
+const defaultStatAllocatedTimers int32 = 128
+
+// NewStat creates and returns a new Stat object.
+func NewStat(statTag string) *Stat {
+	return &Stat{
+		StatTag: statTag,
+		Timers:  make([]Timer, 0, defaultStatAllocatedTimers),
+	}
+}
+
+// NewTimer creates a Timer object within the given stat if stat is non-nil.
+// The value passed in timerTag will be attached to the newly created Timer.
+// NewTimer also automatically sets the Begin value of the Timer to the current
+// time.
+//
+// The user is expected to call stat.Egress with the returned index as argument
+// to mark the end. The return value is not a pointer to a Timer object because
+// the internal slice may be re-allocated freely, which would make the
+// reference to pointers of the past obsolete.
+//
+// Why a slice of timer? To make NewTimer allocation-free, we don't allocate a
+// new timer every time. Instead a slice of timers large enough to hold most
+// data is created and elements from this slice are used. This should put less
+// pressure on the GC too.
+func (stat *Stat) NewTimer(timerTag string) (index int) {
+	if stat != nil {
+		// TODO(adtac): Make this lock-free? We don't expect much contention on
+		// this, so I don't predict a significant improvement in performance. Using
+		// a lock-free, concurrent queue will require us to write our own
+		// reallocation and copy, which is probably unnecessary code complexity if
+		// the improvement isn't significant.
+		stat.mu.Lock()
+		stat.Timers = append(stat.Timers, Timer{
+			TimerTag: timerTag,
+			Begin:    time.Now(),
+			GoId:     goId(),
+		})
+		index = len(stat.Timers) - 1
+		stat.mu.Unlock()
+	}
+
+	return
+}
+
+// Egress marks the completion of a given timer within a stat.
+func (stat *Stat) Egress(index int) {
+	if stat != nil && index < len(stat.Timers) {
+		stat.Timers[index].End = time.Now()
+	}
+}
+
+// AppendTimer appends a given Timer object to the internal slice of timers. A
+// deep copy of the timer is made (i.e. no reference is retained to this
+// pointer) and the user is expected to lose their reference to the timer to
+// allow the Timer object to be garbage collected.
+func (stat *Stat) AppendTimer(timer *Timer) (index int) {
+	if stat != nil && timer != nil {
```

Comment:
Why are you concerned about `stat == nil` in all these methods? When can that happen?

Reply:
If stats are disabled, `http2_server.go` and `http2_client.go` will not create a new `Stat` object at stream initialisation. However, we have calls to stuff like `stream.Stat().NewTimer()` everywhere even though `stream.Stat()` returns nil. This is so that we can have one-line API to hook stats everywhere. That is, adding just one line to the top of a function will automatically do everything:

```go
defer stream.Stat().NewTimer("whatever").Egress()
```

NewTimer will return nil because `stream.Stat()` is nil, and as a result `Egress`, when it is executed at function executed, will also do nothing.

This is simpler than what stat handler and channelz do with ugly `if channelz.IsOn() { ... }` multi-line statements everywhere.

Reply:
See the hooks PR for examples: https://github.com/grpc/grpc-go/pull/3159